File: 鎛維學長的模型/certain.cpp
// 鎛維學長的確定性模型(論文3.2小節)

#include<iostream>
#include <set>
#include <string>
#include"gurobi_c++.h"
using namespace std;

int main()
{
	/*set*/
	set <int> ID;	//確定性航班集合
	set <int> IS;	//不確定性航班集合
	set <int> DD;

	/*params*/
	const int I = 4;	//所有航班數量
	const int K0 = 2;	//所有登機門，0為停機坪
	int A[I] = {400, 500, 600, 700};	//航班i的原訂抵達時間。
	int W = 50;	//更換登機門的懲罰值。
	int B = 100;	//指派至停機坪的懲罰值。
	int S[I] = {30, 45, 60, 50};	//航班i的服務時間。
	int U[I] = {450, 470, 480, 490};	//航班i的更新後抵達時間
	int Y[I][K0] = { {1, 0}, {0, 1}, {1, 0}, {0, 1} };	//航班i指派至登機門k的事前指派。
	int M = 1000;		//不知道這是什麼

	/*gurobi*/
	try
	{
		GRBEnv env = GRBEnv();
		GRBModel model = GRBModel(env);

		string name = "";

		/*variables*/
		GRBVar y[I][K0]; //若 = 1代表航班i使用登機門k
		GRBVar t[I];	//航班i在登機門的開始作業時間。
		GRBVar z[I];	//若=1代表航班i使用之登機門與事前指派不同

		//y[i][k], 若=1代表航班i使用登機門k
		for (int i = 0; i < I; i++)
		{
			for (int j = 0; j < K0; j++)
			{
				name = "y_" + to_string(i) + "_" + to_string(j);
				y[i][j] = model.addVar(0.0, 1.0, 0, GRB_BINARY, name);
			}
		}
		cout << "y sucess" << endl;

		//t[i], 航班i在登機門的開始作業時間。
		for (int i = 0; i < I; i++)
		{
			name = "t_" + to_string(i);
			t[i] = model.addVar(0.0, GRB_INFINITY, 0, GRB_CONTINUOUS, name);
		}
		cout << "t success" << endl;
			
		//z[i], 若=1代表航班i使用之登機門與事前指派不同
		for (int i = 0; i < I; i++)
		{
			name = "z_" + to_string(i);
			z[i] = model.addVar(0.0, 1.0, 0, GRB_BINARY, name);
		}
		cout << "z success" << endl;

		GRBLinExpr sum = 0;


		/*ojective*/
		for (int i = 0; i < I; i++)
			sum += t[i] - A[i] + W*z[i] + B*y[i][0];
			
		model.setObjective(sum, GRB_MINIMIZE);

		/*constraints*/
		//號碼對應到論文
		//(2)
		for (int i = 0; i < I; i++)
		{
			sum = 0;
			for (int k = 0; k < K0; k++)
			{
				sum += y[i][k];
			}
			name = "(2)_" + to_string(i);
			model.addConstr(sum == 1, name);
		}
		cout << "(2) sucess" << endl;
		
		//(3)
		for (int i = 0; i < I; i++)
		{
			for (int j = 0; j < I; j++)
			{
				if (U[i] < U[j])
				{
					for (int k = 0; k < K0; k++)
					{
						name = "(3)_" + to_string(i) + "_" + to_string(j) + "_" + to_string(k);
						model.addConstr(t[i] + S[i] - t[j] <= (2 - y[i][k] - y[j][k]) * M, name);
					}
				}
			}
		}
		cout << "(3) success" << endl;
		
		//(4)
		for (int i = 0; i < I; i++)
		{
			name = "(4)" + to_string(i);
			model.addConstr(U[i] <= t[i], name);
		}
		cout << "(4) success" << endl;
		
		//(5)
		for (int i = 0; i < I; i++)
		{
			sum = 0;
			for (int k = 0; k < K0; k++)
			{
				sum += y[i][k] * (1 - Y[i][k]);
			}
			name = "(5) success" + to_string(i);
			model.addConstr(z[i] >= sum, name);
		}
		cout << "(5) success" << endl;



		model.optimize();

		//輸出y
		cout << "y:" << endl;
		for (int i = 0; i < I; i++)
		{
			for (int j = 0; j < K0; j++)
			{
				if (y[i][j].get(GRB_DoubleAttr_X) == 1)
				{
					cout <<"flight" + to_string(i + 1) + "-->gate " + to_string(j) << endl;
				}
			}
		}

		//輸出t
		cout << "t:" << endl;
		for (int i = 0; i < I; i++)
		{
			cout << "flight" + to_string(i + 1) + ":" << to_string(t[i].get(GRB_DoubleAttr_X)) << endl;
		}

		//輸出z
		cout << "z:" << endl;
		for (int i = 0; i < I; i++)
		{
			cout << "flight" + to_string(i + 1) + ":" << to_string(z[i].get(GRB_DoubleAttr_X)) << endl;
		}
	}
	catch (GRBException e) {
		cout << "Error code = " << e.getErrorCode() << endl;
		cout << e.getMessage() << endl;
	}
	catch (...) {
		cout << "Exception during optimization" << endl;
	}

	system("pause");
	return 0;
}

File: 鎛維學長的模型/certain.py
# 鎛維學長的確定性模型(論文3.2小節)
import gurobipy as gp
from gurobipy import GRB

import gurobipy as gp
from gurobipy import GRB
import pandas as pd
import traceback

# --- Sets ---

df_flights = pd.read_csv("./processed_0623_ver81.csv", encoding="utf-8-sig", dtype=str)
gates = [i for i in range (1, 38, 1)]   # 登機門 K 的集合
apron = [i for i in range (38, 53, 1)]               # 停機坪 (以 0 代表)
locations = apron + gates # 登機門(包含停機坪) K^0 的集合

K = gates
K0 = locations

I = len(df_flights)
flights = []
A=dict()
S=dict()
U=dict()
Y = dict()  # Pre-assigned gates

for _, row in df_flights.iterrows():
  if pd.isna(row["service_time_min"]):
    continue
  flights.append(int(row["flight_index"]))
  A[int(row["flight_index"])] = int(row["start_time_min"])
  S[int(row["flight_index"])] = int(row["service_time_min"])
  U[int(row["flight_index"])] = int(row["updated_start_min"])
  Y[int(row["flight_index"])] = int(row["original_gate_index"])

W = 50  # Penalty for changing gate assignment
B = 100  # Penalty for assigning to apron
M = 100000  # Big-M constant for constraints


# try:
# Create Gurobi environment and model
env = gp.Env()
model = gp.Model("GateAssignment", env=env)

# Variables
# y[i][k]: 1 if flight i is assigned to gate k
y = {}
for i in flights:
    for k in K0:
        y[i,k] = model.addVar(vtype=GRB.BINARY, name=f"y_{i}_{k}")

# t[i]: start time of flight i at gate
t = {}
for i in flights:
    t[i] = model.addVar(vtype=GRB.CONTINUOUS, lb=0.0, name=f"t_{i}")

# z[i]: 1 if flight i's gate differs from pre-assignment
z = {}
for i in flights:
    z[i] = model.addVar(vtype=GRB.BINARY, name=f"z_{i}")

print("Variables created successfully")

# Objective: Minimize total delay + gate change penalties + apron penalties
obj = gp.quicksum(t[i] - A[i] + W*z[i] for i in flights) + B*gp.quicksum(y[i, k] for i in flights for k in apron)
model.setObjective(obj, GRB.MINIMIZE)

# Constraints
# (2)
for i in flights:
    model.addConstr(sum(y[i,k] for k in K0) == 1, name=f"constr_2_{i}")
print("Constraint (2) added")

# (3)
for i in flights:
    for j in flights:
        if U[i] < U[j]:
            for k in K0:
                model.addConstr(
                    t[i] + S[i] - t[j] <= (2 - y[i,k] - y[j,k]) * M,
                    name=f"constr_3_{i}_{j}_{k}"
                )
print("Constraint (3) added")

# (4)
for i in flights:
    model.addConstr(t[i] >= U[i], name=f"constr_4_{i}")
print("Constraint (4) added")

# (5)
for i in flights:
    model.addConstr(
        z[i] >= sum(y[i, k] * (1 - int(Y.get(i, -1) == k)) for k in K0),
        name=f"constr_5_{i}"
    )
print("Constraint (5) added")

# Optimize model
model.optimize()

# Output results
print("\ny:")
for i in flights:
    for k in K0:
        if y[i,k].X > 0.5:  # Check if binary variable is 1
            print(f"flight{i} --> gate {k}")

print("\nt:")
for i in flights:
    print(f"flight{i}: {t[i].X}")

print("\nz:")
for i in flights:
    print(f"flight{i}: {z[i].X}")